src/core/task.ts

```diff
@@ -102,7 +102,13 @@ export class Task<T> {
    * with the given backoff strategy.
    */
   retry(backoff: Backoff, maxRetries: number): Task<T> {
-    return this._retry(backoff, maxRetries, 0);
+    if (maxRetries < 0) {
+      throw new Error('assertion failed: number of task retries must be greater than or equals zero');
```

Comment:
That makes sense. Unfortunately it still means that Task composition would no longer be side-effect free.

Maybe could we find a side-effect free strategy for it instead. Perhaps just don't retry if it's negative? Not sure..